### ISSUE TITLE
Llama-3.1-70B-Q4_1 trace has many slow_memcpy initializer dispatches

### ISSUE THREAD

**Issue by aviator19941** [iree-org/iree#18494]
### What happened?

There are 14,620 child zones that are intializer dispatches of slow_memcpy's. Is there a way to decrease those somehow? The tracy file is provided in Azure [here](https://sharkpublic.blob.core.windows.net/sharkpublic/avi/llama70b_q4_1_prefill_debug.tracy) and the IR is provided in Azure as well [here](https://sharkpublic.blob.core.windows.net/sharkpublic/avi/Llama-3.1-70B-q4_1_v2.mlir).

![Screenshot 2024-09-11 125910](https://github.com/user-attachments/assets/2b64ab26-e1ca-4148-9092-ab10dc286333)

### Steps to reproduce your issue

1. The tracy file is provided in Azure [here](https://sharkpublic.blob.core.windows.net/sharkpublic/avi/llama70b_q4_1_prefill_debug.tracy).
2. Click on the 14,620 `iree_hal_hip_stream_command_buffer` to see the slow_memcpy dispatches.

### Version information

c3cbfbd0cb5e345380deaf8b88938030d30961f9

### Additional context

Related to https://github.com/iree-org/iree/issues/18436

**Comment by aviator19941** [comment]
@IanWood1 
### prefill:
TRACY_NO_EXIT=1 ROCR_VISIBLE_DEVICES=0 ../iree-build-trace/tools/iree-benchmark-module \
  --device=hip://0 \
  --hip_use_streams=true \
  --hip_allow_inline_execution=true \
  --device_allocator=caching \
  --module=./Llama-3.1-70B-q4_1-no_trace.vmfb \
  --parameters=model=../llama70b_q4_1.gguf \
  --function=prefill_bs4 \
  --input=4x16xsi64 \
  --input=4xsi64 \
  --input=4x1xsi64 \
  --input=128x2621440xf16 \
  --benchmark_repetitions=3

### decode:
TRACY_NO_EXIT=1 ROCR_VISIBLE_DEVICES=0 ../iree-build-trace/tools/iree-benchmark-module \
  --device=hip://0 \
  --hip_use_streams=true \
  --hip_allow_inline_execution=true \
  --device_allocator=caching \
  --module=./Llama-3.1-70B-q4_1-no_trace.vmfb \
  --parameters=model=../llama70b_q4_1.gguf \
  --function=decode_bs4 \
  --input=4x1xsi64 \
  --input=4xsi64 \
  --input=4xsi64 \
  --input=4x1xsi64 \
  --input=128x2621440xf16 \
  --benchmark_repetitions=3

**Comment by IanWood1** [comment]
One issue im seeing is a pattern of (see full initializer dispatch [here](https://gist.github.com/IanWood1/535f7c517c284abf57c33929915a5686)):
```mlir
%extracted_slice = tensor.extract_slice %expanded[0, 0, 2] [8192, 256, 8] [1, 1, 1] : tensor<8192x256x10xi16> to tensor<8192x256x8xi16>
%1 = flow.tensor.bitcast %extracted_slice : tensor<8192x256x8xi16> -> tensor<8192x256x16xi8>
```
This will cause the `tensor.extract_slice` to be placed in its own dispatch (causing a slow memcpy) because `DispatchCreation` doesn't handle these types of ops. I think this pattern is coming from https://github.com/iree-org/iree/blob/861695b8b871be9c1ae83bfd786a64a4bb7d5c2f/compiler/plugins/input/Torch/InputConversion/BitCastQuantTensor.cpp#L129-L130


Do we want to be generating `flow.tensor.bitcast` ops when converting from torch? Or does this just need to be added to `DispatchCreation`



For some more context, `flow.tensor.bitcast` seems to come from this in the Torch IR:
```mlir
%__auto.token_embd.weight = util.global.load @__auto.token_embd.weight : tensor<656670720xi8>
%0 = torch_c.from_builtin_tensor %__auto.token_embd.weight : tensor<656670720xi8> -> !torch.vtensor<[656670720],ui8>
%2024 = torch.aten.view.dtype %0, %int2_33 : !torch.vtensor<[656670720],ui8>, !torch.int -> !torch.vtensor<[328335360],si16>
```

**Comment by benvanik** [comment]
nice find - bitcasts should always be fused (never on their own) - they are like reshapes

**Comment by MaheshRavishankar** [comment]
Yeah nice find! Is there a `tensor.bitcast`? Then we could fold those out. Imwe can do it with the flow ops too as part of flow specific canonicalization patterns
